00xx-account-deletion.md

```diff
@@ -0,0 +1,471 @@
+```
+
+Upon removal of `C`, the structure would evolve as follows, with `x` (the empty
+account) marking the freed location.
+
+In this example, locations will be represented as lists of directions for the
+sake of readability. In practice, though, they will be represented by other,
+usually isomorphic to the list of directions, means, e.g., an integer encoding
+the position of the leaf.
+
+The free list now states that the location determined by
+sequence of directions `[Right; Left]` is available for reuse for new
+insertions.
+
+```
+    H(H(A,B),H(C,D))       free = [[Right; Left]]
+      /       \
+   H(A,B)     H(x,D)
+   /  \       /  \
+  A    B     x    D
+
+```
+
+Thus removal incurs a cost proportional to the height of the tree since we have
+to recompute the hashes on the Merkle path from the removed data.
+
+Now adding data `E` would result in
+
+```
+    H(H(A,B),H(E,D))               free = []
+      /       \
+   H(A,B)     H(E,D)
+   /  \       /  \
+  A    B     E    D
+```
+
+##### Merklizing the free list
+
+<!-- The practical usefulness of this section is in discussion  -->
+
+The view of the free list shown in the example above is a simplification.
+
+The data contained in the free list needs to be Merklized as well.  This means,
+assuming a hash function $H$ and locations $loc_1, ..., loc_n$ being inserted in
+the free list in that order, that the actual data in the free list would be:
+
+$$
+free = [ (loc_n, h_n = H(loc_n, h_{n-1})), ..., (loc_2, h_2 = H(loc_2, h_1)); (loc_1, h_1 = H(loc_1, nil))]
+$$
+
+In turn, the ledger state is the Merklized pair of the ledger itself (as a
+Merkle tree) and its free list.
+
+TBD: using an ordered free list to avoid Merklization
+
+
+#### On-disk ledger
+
+The
+[`Database`](https://github.com/MinaProtocol/mina/blob/4495af5caea5e1bb2f98f92592c065f93a586ade/src/lib/merkle_ledger/database.ml)
+module implements an on-disk ledger that backs up the in-memory data structure.
+
+Deletion support relies on the same techniques as in-memory ledgers, that is, a
+representation of a ledger as Merkle tree *and* an additional free list.
+
+The main culprit is the representation of the free list on-disk. Here we will
+piggyback the data on the KV database as follows:
+- the free list is identified by the key `free_list` (like [`last_account_location`](https://github.com/MinaProtocol/mina/blob/4495af5caea5e1bb2f98f92592c065f93a586ade/src/lib/merkle_ledger/database.ml#L235));
+- when parsing binary data associated to this key, we get two things:
+  1. the actual first free location within the Merkle tree;
+  2. the new value of the `free_list`.
+
+Basically, the free list is handled as a binary-encoded list with a lazy decoding function, a single element at a time.
+
+<!-- This needs to be rewritten, not sure this is correct -->
+We will implement a decoding function where one can specify how many locations
+we ideally want to retrieve (up to `n`) to handle functions `set_location_batch`.
```

Comment:
Don't we have to reserialise the whole thing anyway once we've removed these from the list? The data also grows linearly with the number of free'd accounts, can you justify why that's okay and/or propose an alternative?

Reply:
The idea here would have been to deserialize/serialize only once instead of `n` times when do batch allocations.

Reply:
Done